crypto/dilithium/p_dilithium_test.cc

```diff
@@ -679,70 +679,95 @@ TEST(Dilithium3Test, SIGOperations) {
+  const int nid;
+  const size_t public_key_len;
+  const size_t secret_key_len;
+  const size_t signature_len;
+  const char *kat_filename;
+};
 
-  EVP_PKEY_free(dilithium_pkey);
-  EVP_PKEY_CTX_free(dilithium_pkey_ctx);
-  md_ctx.Reset();
-}
+static const struct ML_DSA parameterSet[] = {
+  {"MLDSA67", NID_DILITHIUM3_R3, 1952, 4032, 3309,  "dilithium/kat/mldsa65.txt"},
+};
 
-TEST(Dilithium3Test, KAT) {
-  FileTestGTest("crypto/dilithium/kat/dilithium3_r3.txt", RunKat);
+class MLDSAParameterTest : public testing::TestWithParam<ML_DSA> {};
+
+INSTANTIATE_TEST_SUITE_P(All, MLDSAParameterTest, testing::ValuesIn(parameterSet),
+                         [](const testing::TestParamInfo<ML_DSA> &params)
+                             -> std::string { return params.param.name; });
+
+
+TEST_P(MLDSAParameterTest, KAT) {
+  std::string kat_filepath = "crypto/";
+  kat_filepath += GetParam().kat_filename;
+
+  FileTestGTest(kat_filepath.c_str(), [&](FileTest *t) {
+    std::string count, mlen, smlen;
+    std::vector<uint8_t> seed, msg, pk, sk, sm;
+
+    ASSERT_TRUE(t->GetAttribute(&count, "count"));
+    ASSERT_TRUE(t->GetBytes(&seed, "seed"));
+    ASSERT_TRUE(t->GetAttribute(&mlen, "mlen"));
+    ASSERT_TRUE(t->GetBytes(&msg, "msg"));
+    ASSERT_TRUE(t->GetBytes(&pk, "pk"));
+    ASSERT_TRUE(t->GetBytes(&sk, "sk"));
+    ASSERT_TRUE(t->GetAttribute(&smlen, "smlen"));
+    ASSERT_TRUE(t->GetBytes(&sm, "sm"));
+
+    size_t pk_len = GetParam().public_key_len;
+    size_t sk_len = GetParam().secret_key_len;
+    size_t sig_len = GetParam().signature_len;
+
+    // The KAT files generated by the dilithium team use the optional APIs that
+    // create a signature for a message m and append the message to the end of
+    // the signature. We only want to bring the APIs that create and verify just
+    // the signature, therefore each signature is a constant
+    // DILITHIUM3_SIGNATURE_BYTES and we truncate the signed message down to a
+    // constant DILITHIUM3_SIGNATURE_BYTES.
+
+    std::vector<uint8_t> signature(sig_len);
+    //uint8_t signature[sig_len];
+    sm.resize(sig_len);
+
+    // Convert string read from KAT to int
+    size_t mlen_int = std::stoi(mlen);
+
+    // Here we fix the DRBG (AES-CTR) so that we are able to seed it with the
+    // seed from the KAT (testing only)
+    pq_custom_randombytes_use_deterministic_for_testing();
+    pq_custom_randombytes_init_for_testing(seed.data());
+
+    // Generate our dilithium public and private key pair
+    EVP_PKEY_CTX *dilithium_pkey_ctx = EVP_PKEY_CTX_new_id(GetParam().nid, nullptr);
+    ASSERT_NE(dilithium_pkey_ctx, nullptr);
+
+    EVP_PKEY *dilithium_pkey = EVP_PKEY_new();
+    ASSERT_NE(dilithium_pkey, nullptr);
+
+    EXPECT_TRUE(EVP_PKEY_keygen_init(dilithium_pkey_ctx));
+    EXPECT_TRUE(EVP_PKEY_keygen(dilithium_pkey_ctx, &dilithium_pkey));
+    const DILITHIUM3_KEY *dilithium3Key = (DILITHIUM3_KEY *)(dilithium_pkey->pkey.ptr); //called dilithium3 but is generic
```

Comment:
Renaming everything to ML-DSA lingo at some point?

Reply:
Yes, will be switching over to ML-DSA lingo in subsequent PRs (including real OIDs). It makes more sense to do this when I add the other 2 parameter sets for ML-DSA, to keep this PR focused on just one addition.